File: hayeong/2023_10/1012/PGS_연속된부분수열의합.java
import java.util.Arrays;
import java.util.PriorityQueue;

public class PGS_연속된부분수열의합 {
	public static int[] solution(int[] sequence, int k) {
		PriorityQueue<Point> pq = new PriorityQueue<>(
				(p1, p2) -> p1.len == p2.len ? p1.start - p2.start : p1.len - p2.len);
		int start = 0;
		int end = 0;
		int sum = sequence[0];
		while (end < sequence.length && start <= end) {
			if (sum < k) {
				end++;
				if(end<sequence.length)
					sum += sequence[end];

			} else if (sum > k) {
				sum -= sequence[start];
				start++;
			} else {
				pq.add(new Point(start, end, end - start + 1));
				sum -= sequence[start];
				start++;
				end++;
				if(end<sequence.length)
					sum += sequence[end];
				
			}
		}

		Point result = pq.poll();
		int[] answer = {result.start, result.end};
		return answer;
	}

  
	static class Point {
		int start, end, len;

		Point(int start, int end, int len) {
			this.start = start;
			this.end = end;
			this.len = len;
		}
	}
}

File: woojae/2024_5/0512/PGS_연속된 부분 수열의 합.py
def solution(sequence, k):
    low = high = 0
    answer = [0, len(sequence)]  # 전체 범위
    target = sequence[0]

    while True:
        if target < k:
            high += 1
            if high == len(sequence):  # high가 배열 범위를 벗어났을 경우
                break
            target += sequence[high]  # 새로운 값이 추가됐으므로 값 합산
        else:  # 타겟과 같거나 작을 때
            if target == k:  # 타겟을 찾았다면
                if high - low < answer[1] - answer[0]:  # 또한 현재 수열보다 길이가 짧다면
                    answer = [low, high]  # 답 갱신
            target -= sequence[low]  # 필요없는 값이 됐으므로 제거 
            low += 1
            
    return answer
